commit 18404ea60141a2e2eaf58a5ebbd2b99f7a0cd442
Author: Masahiko Sawada <msawada@postgresql.org>
Date:   Wed Jun 12 09:56:13 2024 +0900

    Fix segmentation fault in test_tidstore.
    
    The do_set_block_offsets() and other functions accessing the tidstore
    did not check if the tidstore was NULL. This led to a segmentation
    fault when these functions are called without calling the
    test_create().
    
    This commit adds NULL checks in relevant functions of test_tidstore to
    raise an error instead if the tidstore is not initialized.
    
    Bug: #18483
    Reported-by: Alexander Kozhemyakin
    Reviewed-by: Michael Paquier
    Discussion: https://postgr.es/m/18483-30bfff42de238000%40postgresql.org

diff --git a/src/test/modules/test_tidstore/test_tidstore.c b/src/test/modules/test_tidstore/test_tidstore.c
index 5417163407b..3f6a11bf21c 100644
--- a/src/test/modules/test_tidstore/test_tidstore.c
+++ b/src/test/modules/test_tidstore/test_tidstore.c
@@ -144,10 +144,17 @@ sanity_check_array(ArrayType *ta)
 		ereport(ERROR,
 				(errcode(ERRCODE_DATA_EXCEPTION),
 				 errmsg("argument must be empty or one-dimensional array")));
 }
 
+static void
+check_tidstore_available(void)
+{
+	if (tidstore == NULL)
+		elog(ERROR, "tidstore is not created");
+}
+
 static void
 purge_from_verification_array(BlockNumber blkno)
 {
 	int			dst = 0;
 
@@ -165,10 +172,11 @@ do_set_block_offsets(PG_FUNCTION_ARGS)
 	BlockNumber blkno = PG_GETARG_INT64(0);
 	ArrayType  *ta = PG_GETARG_ARRAYTYPE_P_COPY(1);
 	OffsetNumber *offs;
 	int			noffs;
 
+	check_tidstore_available();
 	sanity_check_array(ta);
 
 	noffs = ArrayGetNItems(ARR_NDIM(ta), ARR_DIMS(ta));
 	offs = ((OffsetNumber *) ARR_DATA_PTR(ta));
 
@@ -215,10 +223,12 @@ check_set_block_offsets(PG_FUNCTION_ARGS)
 	TidStoreIterResult *iter_result;
 	int			num_iter_tids = 0;
 	int			num_lookup_tids = 0;
 	BlockNumber prevblkno = 0;
 
+	check_tidstore_available();
+
 	/* lookup each member in the verification array */
 	for (int i = 0; i < items.num_tids; i++)
 		if (!TidStoreIsMember(tidstore, &items.insert_tids[i]))
 			elog(ERROR, "missing TID with block %u, offset %u",
 				 ItemPointerGetBlockNumber(&items.insert_tids[i]),
@@ -303,19 +313,23 @@ check_set_block_offsets(PG_FUNCTION_ARGS)
 Datum
 test_is_full(PG_FUNCTION_ARGS)
 {
 	bool		is_full;
 
+	check_tidstore_available();
+
 	is_full = (TidStoreMemoryUsage(tidstore) > tidstore_empty_size);
 
 	PG_RETURN_BOOL(is_full);
 }
 
 /* Free the tidstore */
 Datum
 test_destroy(PG_FUNCTION_ARGS)
 {
+	check_tidstore_available();
+
 	TidStoreDestroy(tidstore);
 	tidstore = NULL;
 	items.num_tids = 0;
 	pfree(items.insert_tids);
 	pfree(items.lookup_tids);
